### PR TITLE
Keys readme file

### PR DESCRIPTION
**Pull request by maririos**

### CODE REVIEW

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+using Azure.Identity;
+using Azure.Security.KeyVault.Keys;
+
+// Create a new key client using the default credential from Azure.Identity
+var client = new KeyClient(vaultUri: <your-vault-url>, credential: new DefaultAzureCredential());
+
+// Create a new key using the key client
+Key key = await Client.CreateKey("key-name", KeyType.EllipticCurve);
+```
+> DefaultAzureCredential():
```
Comment (tg-msft) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: Maybe make it `new DefaultAzureCredential()` or just `DefaultAzureCredential`?

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+> Uses the environment variables previously set (`AZURE_CLIENT_ID`, `AZURE_CLIENT_SECRET`, and `AZURE_TENANT_ID`).
+
+## Key concepts
+### Keys
+Azure Key Vault supports multiple key types and algorithms, and enables the use of Hardware Security Modules (HSM) for high value keys.
+
+### Key Client:
+An asynchronous and synchronous KeyClient client exists in the SDK allowing for selection of a client based on an application's use case. Once you've initialized a KeyClient, you can interact with the primary resource types in Key Vault.
```
Comment (tg-msft) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: > An asynchronous and synchronous KeyClient client exists

Are there two separate clients?  If not, maybe reword to

> A `KeyClient` providing both synchronous and asynchronous operations exists

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+### Sync examples
+* [Create a Key](#create-a-key)
+* [Retrieve a Key](#retrieve-a-key)
+* [Update an existing Key](#update-an-existing-key)
+* [Delete a Key](#delete-a-key)
+* [List Keys](#list-keys)
+
+### Async examples
+* [Create a Key](#async-create-a-key)
```
Comment (tg-msft) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: Is this the balance we want between sync/async in the README?  For Storage I was planning on flipping this with one sync example and a handful of async examples.  I can adjust to match this if the @kaerm thinks it's more helpful.

**Review by kaerm — APPROVED** [review]

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+### Create a Key
+Create a Key to be stored in the Azure Key Vault. If a key with the same name already exists, then a new version of the key is created.
+
+```c#
+// Create a key of any type
```
Comment (tg-msft) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: What does 'of any type' mean here?  All the other examples call out the type of the key being created in the comment like `// Create a software Elliptic Curve key`

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+Key key = client.CreateKey("key-name", KeyType.EllipticCurve);
+
+Console.WriteLine(key.Name);
+Console.WriteLine(key.KeyMaterial.KeyType);
```
Comment (tg-msft) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: Nit - maybe consider merging these into a single line like `Console.WriteLine($"Key {key.Name} has type {key.KeyMaterial.KeyType}");` so the examples are slightly shorter.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+Console.WriteLine(key.Name);
+Console.WriteLine(key.DeletedDate);
+```
+
+### List Keys
+This example lists all the keys in the specified Key Vault.
+
+```c#
+List<Response<KeyBase>> allKeys = Client.GetKeys();
+
```
Comment (tg-msft) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: This line looks suspect.  Is it `List<Response<KeyBase>>` or `Response<List<KeyBase>>`?  And should `Client` be lowercase?

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+  foreach (Key key in allKeys)
+  {
+    Console.WriteLine(key.Name);
+  }
+```
+
+### Async create a Key
+Async APIs are identical to synchronous API. Note that all methods end with `Async`.
```
Comment (tg-msft) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: Maybe reword to

> Async APIs are identical to their synchronous counterparts.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+### Retrieve a Key
+`GetKey` retrieves a key previously stored in the Key Vault.
+
+```c#
+Key key = client.GetKey("key-name");
+
+Console.WriteLine(key.Name);
+Console.WriteLine(key.KeyMaterial.KeyType);
+```
+
```
Comment (tg-msft) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: Maybe add a comment linking to the errors section so they can see what to do about a key not existing?

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+## Troubleshooting
+
+### General
+When you interact with the Azure Key Vault Key client library using the .NET SDK, errors returned by the service correspond to the same HTTP status codes returned for [REST API][keyvault_rest] requests.
+
+For example, if you try to retrieve a Key that doesn't exist in your Key Vault, a `404` error is returned, indicating `Not Found`.
+
+```c#
+Key key = client.GetKey("some_key");
```
Comment (tg-msft) [inline sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: You should update this to wrap in a `try/catch`, show the type of exception, and `Console.WriteLine(ex.ToString())` to help users get started.

**Review by tg-msft — APPROVED** [review]
Looks great.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md [hunk 10]
```diff
@@ -0,0 +1,282 @@
+# Azure Key Vault Key client library for .NET
+Azure Key Vault is a cloud service that provides a secure storage of keys. Multiple keys, and multiple versions of the same key, can be kept in the Key Vault. Cryptographic keys in Key Vault are represented as [JSON Web Key (JWK)][JWK] objects.
```
[the remark this answers is not in the document]
Comment (kaerm) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: good point

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md (hunk 3 above)
Comment (maririos) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: That's how we've been doing it in other readmes for KV, and what I understood from Kate. Happy to change in case it is wrong

Comment (kaerm) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: In general I am all for equal representation of sync and async but I think different languages took different approaches to that. From what I understand the main focus was on providing sync libraries first as they are easier to understand for people getting started and then async because people who know they need async will choose it anyway. In general, I would provide all examples for both of them (with sync listed first but clear mention that async exists) but if the samples are very similar and just need an additional word then I am fine with the way it is listed here. Anyway, that's what I would suggest and we can verify if it gets confusing for people to use

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md [hunk 11]
```diff
@@ -0,0 +1,282 @@
+# Azure Key Vault Key client library for .NET
+Azure Key Vault is a cloud service that provides a secure storage of keys. Multiple keys, and multiple versions of the same key, can be kept in the Key Vault. Cryptographic keys in Key Vault are represented as [JSON Web Key (JWK)][JWK] objects. 
+
+The Azure Key Vault keys library client supports RSA keys and Elliptic Curve (EC) keys, each with corresponding support in hardware security modules (HSM). It offers operations to create, retrieve, update, delete, purge, backup, restore and list the keys and its versions.
+
+[Source code][keys_client_src] | Package (NuGet) (coming soon) | [API reference documentation] (coming soon) | [Product documentation][keyvault_docs] | [Samples] (coming soon)
```
[the remark this answers is not in the document]
Comment (kaerm) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: Good point - maybe at the bottom in the additional documentation section? something along the lines of "If you are looking to do x click here"

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md (hunk 6 above)
Comment (maririos) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: GetKeys returns `IAsyncEnumerable<Response<KeyBase>>`

Comment (tg-msft) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: Ah, since that's the sync API, should it be regular `IEnumerable` as well?

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (maririos) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: Hmm not sure if we want to link the different packages in the readme. So far we've kept them separated.

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+
+## Getting started
+
+### Install the package
+Install the Azure Key Vault Keys client library for .NET with [NuGet][nuget]:
+
+```PowerShell
+Install-Package Azure.Security.KeyVault.Keys -Version 1.0.0-preview.1
```
[the remark this answers is not in the document]
Comment (maririos) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: Nice. Thank you!

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md (hunk 6 above)
Comment (maririos) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: yeah you are right, it was `IEnumerable<Response<KeyBase>>` and yeah, that is better. I'll change it

File: sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (maririos) [reply sdk/keyvault/Azure.Security.KeyVault.Keys/Readme.md]: Added in additional documents section
